src/main/java/io/supertokens/inmemorydb/Start.java

```diff
@@ -1619,4 +1627,123 @@ public void addInfoToNonAuthRecipesBasedOnUserId(String className, String userId
+        try {
+            int updatedCount = TOTPQueries.markDeviceAsVerified(this, userId, deviceName);
+            if (updatedCount == 0) {
+                throw new UnknownDeviceException();
+            }
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void deleteDevice(String userId, String deviceName)
+            throws StorageQueryException, UnknownDeviceException {
+        try {
+            int deletedCount = TOTPQueries.deleteDevice(this, userId, deviceName);
+            if (deletedCount == 0) {
+                throw new UnknownDeviceException();
+            }
+
+            // FIXME: Should delete all the codes associated with this device.
+            // But problem is that we store all the codes whether they are valid (i.e.
+            // matching device found) or not.
+            // So we can't just delete all the codes for the user. We need to delete only
+            // the codes that are associated
+            // with this device. But we don't store that information.
+
+            // One way is only delete the valid codes?
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public void updateDeviceName(String userId, String oldDeviceName, String newDeviceName)
+            throws StorageQueryException, DeviceAlreadyExistsException,
+            UnknownDeviceException {
+        try {
+            int updatedCount = TOTPQueries.updateDeviceName(this, userId, oldDeviceName, newDeviceName);
+            if (updatedCount == 0) {
+                throw new UnknownDeviceException();
+            }
+        } catch (SQLException e) {
+            if (e.getMessage()
+                    .equals("[SQLITE_CONSTRAINT]  Abort due to constraint violation (UNIQUE constraint failed: "
+                            + Config.getConfig(this).getTotpUserDevicesTable() + ".user_id, "
+                            + Config.getConfig(this).getTotpUserDevicesTable() + ".device_name" + ")")) {
+                throw new DeviceAlreadyExistsException();
+            }
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public TOTPDevice[] getDevices(String userId)
+            throws StorageQueryException {
+        try {
+            return TOTPQueries.getDevices(this, userId);
+        } catch (SQLException e) {
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Override
+    public boolean insertUsedCode(TOTPUsedCode usedCodeObj)
+            throws StorageQueryException, TotpNotEnabledException {
+        try {
+            TOTPDevice[] devices = TOTPQueries.getDevices(this, usedCodeObj.userId);
+            if (devices.length == 0) {
+                throw new TotpNotEnabledException();
+            }
+
+            int insertCount = TOTPQueries.insertUsedCode(this, usedCodeObj);
+            return insertCount == 1;
```

Comment:
- this is not correct - cause the insert query that you wrote will fail and not return 0.
- Add a test for this

Reply:
Fixed. Also, emutated foreign key constraints now in SQL Transaction.